2-mandatory/8-password-validator.js

```diff
@@ -23,7 +23,24 @@ PasswordValidationResult=  [false, false, false, false, true]
 
 */
 
-function validatePasswords(passwords) {}
+function validatePasswords(passwords) {
+  const prPassword=[];
+  return passwords.map((password)=>{
+    if(password.length>4 && 
+    // /[A-Z]/g.test(password) &&
+    containsUppercaseLetter(password)&&
+    // /[a-z]/g.test(password) &&
+    containsLowercaseLetter(password)&&
+  //  /\d/g.test(password)&&
+  containsNumber(password)&&
+  //  /[!#$%.*&]/g.test(password)&&
+  containsSymbol(password) &&
+   !prPassword.includes(password)){
+      prPassword.push(password);
+      return true;
+    }else return false;
+  });
+}
 
```

Comment:
I see you're saving previous passwords. Here's my solution:
```javascript
function validatePasswords(passwords) {
  return passwords.map((password, index) => {
    // - Have at least 5 characters.
    if (password.length < 5) return false;
    // - Have at least one English uppercase letter (A-Z)
    if (!containsUppercaseLetter(password)) return false;
    // - Have at least one English lowercase letter (a-z)
    if (!containsLowercaseLetter(password)) return false;
    // - Have at least one number (0-9)
    if (!containsNumber(password)) return false;
    // - Have at least one non-alphanumeric symbol ("!", "#", "$", "%", ".", "*", "&")
    if (!containsSymbol(password)) return false;
    // - Must not be any previous password in the passwords array.
    if (passwords.slice(0, index).find((pwd) => pwd === password)) return false;
    // Must be OK!
    return true;
  });
}
```
In this version, I'm using the principle of _early exit_ from my _map_ predicate. Rather than string all the password rules together into one giant condition, I have a condition per rule and return _false_ as soon as a rule is broken. This is much easier to debug than a giant condition. I've also left the rule comments in place, as you did too, so I can be sure I have not forgotten a rule.

The checking of the "cannot be a previous password" rule is the hardest part of this exercise. Rather than keep a list of previously processed passwords, I just used slice to get the passwords from the start up to the _current_ password. I didn't test this thoroughly though so it could be a naive implementation.